src/modules/context2d.js

```diff
@@ -2037,12 +2070,24 @@ import { console } from "../libs/console.js";
           oldLineWidth = this.lineWidth;
           this.lineWidth = oldLineWidth * options.scale;
         }
-        this.pdf.text(options.text, tmpRect.x, tmpRect.y, {
-          angle: options.angle,
-          align: textAlign,
-          renderingMode: options.renderingMode,
-          maxWidth: options.maxWidth
-        });
+
+        if (tmpRect.y <= pageHeightMinusMargin) {
+          if (tmpRect.y - tmpRect.h >= 0 && tmpRect.x <= pageWidthMinusMargin) {
+            var croppedText = this.pdf.splitTextToSize(options.text, options.maxWidth || pageWidthMinusMargin - tmpRect.x)[0];
+            this.pdf.text(croppedText, tmpRect.x, tmpRect.y, {
+              angle: options.angle,
+              align: textAlign,
+              renderingMode: options.renderingMode,
+              renderMaxWidthOverflow: false
+            });
+          }
+        } else {
+          // This text is the last element of the page, but it got cut off due to the margin
+          // so we render it in the next page
+
+          // As a result, all other elements have their y offset increased
+          this.ctx.prevPageLastElemOffset += pageHeightMinusMargin - tmpRect.y + tmpRect.h;
+        }
```

Comment:
My previous comment regarding this code:

> Both if conditions are not needed, they can be removed.
This is because getPagesByPath is already accounting for margins.
Also this code was falsely accusing out of bonds characters and when this happened, these characters would not be drawn(because the for loop through the pages only iterates once because there only is one page).

Was wrong, there are certain situations where one character can be attributed 2 pages (right on the edge of the bottom margin). The code was wrongfully pointing out characters as being out of range when they were not however. This is because the comparison: 
`tmpRect.y <= pageHeightMinusMargin` is not correct. tmpRect.y is a value that can range from 0 to the page height. pageHeightMinusMargin can range from 0 to page height minus the margins. We're interested in knowing whether or not a character fits in the page boundaries delimited by the margins and this code is not doing that.

```suggestion

       if ((this.margin[0] <= tmpRect.y) && (tmpRect.y <= (this.pdf.internal.pageSize.height - this.margin[2]))) {
              var croppedText = this.pdf.splitTextToSize(options.text, options.maxWidth || pageWidthMinusMargin - tmpRect.x)[0];
              this.pdf.text(croppedText, tmpRect.x, tmpRect.y, {
                  angle: options.angle,
                  align: textAlign,
                  renderingMode: options.renderingMode,
                  renderMaxWidthOverflow: false
              });
          }
```

Reply:
To me, it looks like there is something severely broken. If you take the `page break with text` test case in the `html.spec`, the margins don't arrive in the context2d module at all and the the `tmpRect.y` coordinate for the first `putText` call with "Lorem " is completely wrong (its ~846, but should be something closer to 0, the y coordinate for the next text part "ipsum" is 33.9).

If I understand correctly, `putText` will always draw only a single line of text. So the `pages` array has either only one page, if the line bounds fit entirely on the current page, or two pages if the line bounds don't fit entirely. If the pages array has two pages, we should simply draw the text on the second page. In this case we also need to shift subsequent text further down.